### PR TITLE
Feature/User Search

### PR DESCRIPTION
**Pull request by shepherd-06**
#Feature 
Search user by firstname, lastname, email or username.

Returns a list a of users in this format. 
```
[
  {
    "id": "string",
    "email": "string",
    "userName": "string",
    "firstName": "string",
    "lastName": "string",
  }
]
```
Must merge after #53

### CODE REVIEW

File: backend/server/src/models/tournamentModel.ts [hunk 1]
```diff
@@ -39,12 +68,14 @@ const tournamentSchema = new Schema<Tournament & Document>(
     },
     maxPlayers: { type: Number, required: true },
     players: [{ type: String }],
-    matchSchedule: [{ type: Schema.Types.ObjectId, ref: "Match" }], // Reference to Match documents
+    matchSchedule: [{ type: String }], // Reference to Match documents
```
Comment (vilbedawg) [inline backend/server/src/models/tournamentModel.ts:71]: Why is this a string type?

File: backend/server/src/models/tournamentModel.ts [hunk 2]
```diff
@@ -1,12 +1,39 @@
...
+  playersDetails: Array<{
+    id: string | null;
+    firstName: string | null;
+    lastName: string | null;
+  }>;
+  winnerDetails?: {
+    id: string | null;
+    firstName: string | null;
+    lastName: string | null;
+  } | null;
```
Comment (vilbedawg) [inline backend/server/src/models/tournamentModel.ts:34]: Do the properties need to be nullable as well considering that players are required to enter all of these information, therefore if the player exists, then these would exist as well?

File: backend/server/src/models/tournamentModel.ts [hunk 3]
```diff
@@ -1,12 +1,39 @@
...
+  type: string;
+  admin: Types.ObjectId | null;
+  elapsedTime: number;
+  timerStartedTimestamp: Date | null;
+}
+
+interface PlayerDetails {
+  firstName: string | null;
+  lastName: string | null;
+  id: string | null;
```
Comment (vilbedawg) [inline backend/server/src/models/tournamentModel.ts:21]: Same here

File: backend/server/src/controllers/tournamentController.ts [hunk 4]
```diff
@@ -23,6 +27,16 @@ export class TournamentController extends Controller {
     return await this.service.getTournamentById(id);
   }
 
+  @Security("jwt")
+  @Get("{id}/details")
```
Comment (vilbedawg) [inline backend/server/src/controllers/tournamentController.ts:31]: I think that specifying this endpoint as just `{id}` is sufficient and more REST -like

File: backend/server/src/services/userService.ts [hunk 5]
```diff
@@ -34,4 +34,16 @@ export class UserService {
...
+  public async searchUser(query: string): Promise<User[]> {
+    const searchQuery = new RegExp(query, 'i');
+    return await UserModel.find({
+      $or: [
+        { userName: { $regex: searchQuery } },
+        { email: { $regex: searchQuery } },
+        { firstName: { $regex: searchQuery } },
+        { lastName: { $regex: searchQuery } }
+      ]
+    });
```
Comment (vilbedawg) [inline backend/server/src/services/userService.ts]: These should be converted into plain objects before sending them to the client. I think now if you make a request, you will get the document data as well, which is not what we want.

**Comment by vilbedawg** [comment]
Should we close this @shepherd-06 since we don't have time to fully implement this?

**Comment by shepherd-06** [comment]
I am closing this PR. It's best to do further development on this feature from scratch.

**Pull request closed without merging**